### DISCUSSION TITLE
Patterns explained

### DISCUSSION THREAD

**Discussion by esseti** [danielmiessler/Fabric#1091] [Ideas]
Is there a website or similar platform where one can search for and list what different patterns do, ideally featuring a curated list with examples of how to use them effectively?

**Comment by yousecjoe** [comment]
The patterns are all markdown, which you can read with any text editor to see what they do. 

There is also an example pattern you can review. 

I could create what you're asking for, but I'm not sure if it is worth the time and is necessary when the patterns are all self-explanatory. 

The source repo can be used for the use case you describe. Perhaps I'm mistaken. 

There are also YT videos from the creator that showcase some of the patterns. 

What is it you're looking for that isn't available using GitHub?

**Reply by esseti** [reply]
You must go through the list of all the patterns in the repository to understand their functions. A Readme.md that summarizes each pattern in one paragraph would likely suffice to display this information with the `-l` command. This way, even someone who has not read all of them can understand what they do.

**Reply by yousecjoe** [reply]
> You must go through the list of all the patterns in the repository to understand their functions.

Yes, that's going to be required to use the software effectively. "Exploring" is recommended in the main repo README.md. 

> A Readme.md that summarizes each pattern in one paragraph would likely suffice to display this information with the `-l` command. This way, even someone who has not read all of them can understand what they do.

I agree there needs to be better documentation for many of the prompts. I couldn't begin to tell you how many exactly. Perhaps you have better data. 

There are some things I'm curious about. For example, how do we overcome the lack of an official or standard structure to develop a prompt?

This means someone will have to maintain and validate all these README.md files every time a prompt changes. Could it be done, yes? Would it be efficient and effective? That's what I'm curious about.  

Whoever maintains the README.md files would need to keep up with the pace of new prompts developed, while the prompt author has no requirement to establish the README.md, there is only going to be a SYSTEM.md, which often includes the `IDENTITY AND PURPOSE`, which is most often all you need to read to get an idea of what a prompt does. Most prompt README.md files would be duplicating the `IDENTITY AND PURPOSE` line of the SYSTEM.md file. That seems unnecessary to do, my 2c. 

From my experience, the issue arises when people develop prompts that do not follow the `IDENTITY AND PURPOSE` template. As long as all prompt templates are allowed into the pattern library, it will be interesting to see how to document them. 

Perhaps using Fabric's ability to summarize the patterns is one way to keep things somewhat coherent, concise, and up to date. 

However, I can see the prompt `analyze_answers` has a README.md, which works well to explain how it is explicitly used. Someone intimate with that prompt's functionality can only achieve that level of detail. Unless I'm mistaken, this means someone must be intimate with every prompt that is developed to have the ability to write a README.md.

**Reply by esseti** [reply]
One thing could be that:
`  -U, --updatepatterns      Update patterns`
is removed as parameter function and moved to the setup phase (which is way more intuitive to have it there)

then the command `fabric install updatepatterns --ai-description` download all the patterns and for each folder, that does not contains a readme.md file, it calls a prompt (that has to be created, but does not seems too complex) which has the goal of creating the readme.md file starting with all the content in the folder. So that each folder at the end of the operation will have a readme that describes the pattern. 
This readme.md is then printed in the command `fabric -l` 

Alternatively it could be forced when merging into main bracnh has to check that all the folders have their own readme. but probably a bit too strickt?

**Comment by noamsiegel** [comment]
I am working on a fabric front end. It has a page that shows all the patterns. It would be a good idea to add a page that shows all the fabrics and their descriptions...
